hashes/src/sha256.rs

```diff
@@ -114,31 +112,12 @@ impl Hash {
 #[derive(Copy, Clone, PartialEq, Eq, Default, PartialOrd, Ord, Hash)]
 pub struct Midstate(pub [u8; 32]);
 
-crate::internal_macros::arr_newtype_fmt_impl!(Midstate, 32);
-serde_impl!(Midstate, 32);
-borrow_slice_impl!(Midstate);
-
-impl<I: SliceIndex<[u8]>> Index<I> for Midstate {
-    type Output = I::Output;
-
-    #[inline]
-    fn index(&self, index: I) -> &Self::Output { &self.0[index] }
-}
-
-impl core::str::FromStr for Midstate {
-    type Err = hex::HexToArrayError;
-    fn from_str(s: &str) -> Result<Self, Self::Err> { hex::FromHex::from_hex(s) }
-}
+crate::impl_bytelike_traits!(Midstate, 32, true);
```

Comment:
Isn't midstate mainly intended for creating tags? It thus shouldn't need all the traits.

But even more seriously, didn't we agree to have length in the midstate? That'd be incompatible with serialization. Do we really want to do this?

Reply:
Good points. I think we should back up, make the `Midstate` be a struct with a separate length and byte array (and maybe even a byte array for unhashed bytes? but then it would be identical to the `Engine` type)

We should provide inherents to access both fields, as well as a convenience function to check whether the length is a full multiple of 64 (or maybe this should be an invariant, and when extracting a midstate from an engine that has a non-multiple-of-64 it should fail?)

I think maybe we need to promote this to its own issue/discussion. Meanwhile I think the existing PR is API-diff-minimizing so we can accept it for now.

Reply:
Promoted to https://github.com/rust-bitcoin/rust-bitcoin/issues/2918

I'll leave it up to you whether you want to resolve that before accepting this PR.

Reply:
I think we should resolve it before accepting this specific change. The rest of the PR looks fine. (And no, I really don't want to add API that we suspect we will to remove in a subsequent PR potentially in a next release. Too much noise.)